### ISSUE TITLE
SQLite : error Error: SQLITE_ERROR: too many SQL variables

### ISSUE THREAD

**Issue by Vinnythetrue** [strapi/strapi#7214]
Hi,

I'm using Strapi on a project as a backend api, globally I'm greatly satisfied, even though lately, without any warnings, it suddenly started responding with an error on the admin panel for one of the entities, and only this one.
The structure has not lately changed, the only this that has is that data has been added in this very entity.

Pastebin to these structures : https://pastebin.com/B0UQxw04

Here is a view of the model :
```json
{
  "kind": "collectionType",
  "collectionName": "forms",
  "info": {
    "name": "form"
  },
  "options": {
    "increments": true,
    "timestamps": true
  },
  "attributes": {
    "title": {
      "type": "string"
    },
    "questions": {
      "type": "component",
      "repeatable": true,
      "component": "form.question"
    },
    "submission_message": {
      "type": "string"
    },
    "instructions": {
      "type": "text"
    }
  }
}
```

As you see, it uses the component question :
```json
{
  "collectionName": "components_form_questions",
  "info": {
    "name": "question",
    "icon": "question"
  },
  "options": {},
  "attributes": {
    "category_id": {
      "model": "category"
    },
    "text": {
      "type": "string"
    },
    "multiple_allowed": {
      "type": "boolean",
      "default": false
    },
    "responses": {
      "type": "component",
      "repeatable": true,
      "component": "form.response"
    }
  }
}
```

Which uses the component response :
```json
{
  "collectionName": "components_form_responses",
  "info": {
    "name": "response",
    "icon": "award"
  },
  "options": {},
  "attributes": {
    "text": {
      "type": "string"
    },
    "points": {
      "type": "component",
      "repeatable": true,
      "component": "form.point"
    }
  }
}
```

and finally this one uses the component point :
```json
{
  "collectionName": "components_form_responses",
  "info": {
    "name": "response",
    "icon": "award"
  },
  "options": {},
  "attributes": {
    "text": {
      "type": "string"
    },
    "points": {
      "type": "component",
      "repeatable": true,
      "component": "form.point"
    }
  }
}
```

There are only 2 form entities existing in the database, but I have to say with something like 40 questions each, with each 6-7 responses each with 1-3 points. Deploying a new instance with exact same code works very well. So it seems really to be a data-related issue. I cannot really see what I could try from that point to be honest.

Vinny

**Comment by alexandrebodin** [comment]
Hey @Vinnythetrue can you use the markdown code block to make your issue a bit easier to read ? :) Thanks !

**Comment by Vinnythetrue** [comment]
@alexandrebodin OK, I tried doing this, not quite convinced by the result though... I added a pastebin link, hope it's ok this way

**Comment by alexandrebodin** [comment]
@Vinnythetrue Updated it to make easier to read no worries :)

**Comment by alexandrebodin** [comment]
Are you getting the error when you load the list view or when you load a specific content ? any search or filter ?

**Comment by Vinnythetrue** [comment]
I thought the list view through the API was working, but it's actually not.
So the api route /forms without any filter/whatsoever is broken.
But getting either /forms/3 or /forms/4 is working pretty well.

I am just checking the data, 120 questions total (40 for form 3, 80 for form 4), only 2 form entities.

If I manually delete the form 4 in my database, everything's back to normal. I'm going to check what happens if I remove questions, if maybe there is some kind of variable cound limitation I have reached ? I read things about 999 variables. Is there a way to increase this to check ? Through database.js maybe ?

**Comment by alexandrebodin** [comment]
Actually I have no idea it's the first time this every happens. If sqlite is too limiting I would suggestion you move th Postgres or Mysql for production

**Comment by Vinnythetrue** [comment]
OK, I see what is happening... First a select forms; is issued. Then a select for the found form_id (3,4 at this point), to retrieve the form components (all the questions, let's say 120 approx.), then another sleect to retrieve all the responses, and finally one last to retrieve all (more 900 !!!) points subsubsubcomponents...

Therefore issuing this error, because each component id is a bound parameter.:

```
  sql:
   'select `components_form_points`.* from `components_form_points` where `id` in (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' }
```

**Comment by alexandrebodin** [comment]
Wow ok so that is a limitation of SQLITE itslef. it would require to load the components manually instead of using the autoloading feature of bookshelf. This is not handled in bookshelf itself https://github.com/bookshelf/bookshelf/issues/707

There isn't much workaround except changin DB or changing your data strucutre :/

**Comment by Vinnythetrue** [comment]
Yeah, I figured that one out :(
I think changing DB would right now be my best option, but what will happen in the future ? From what I see, this way of generating subselects with more and more id lists doesn't really make sense, it would need joins or something.
Maybe loading manually like you're talking about ? I assume I may find guides on customising strapi controllers code ? Even though I did chose strapi in the first place because time to develop is an issue on that project :(

**Comment by alexandrebodin** [comment]
This should actually be handled in the bookshelf project so it splits the queries into chunks to avoid this

**Comment by animatedcreativity** [comment]
I got into same problem when doing an API request such as: `/categories?_start=800&_limit=100`
It gave me same error in the case: `error Error: SQLITE_ERROR: too many SQL variables`
So, I tried to lower the limit to 20 and it worked: `/categories?_start=800&_limit=20`

Reducing the limit fetches data slower for me now for like 1500 records, but I am happy that it works now.
Hope it helps someone.

Cheers!

**Comment by derrickmehaffy** [comment]
Going to mark the issue as closed as no new information has been added.

**Comment by wysher** [comment]
I encountered simillar problem. In my case I have relatively list of participants (> 1000 entities). But... because of complicated filtering I cannot use query mechanism, and I do:

```
const found = await strapi.services.participant.find({ _limit: -1 });
const result = found.filter(...)
```
I know it's not ideal, but must be this way.

With this I get `SQLITE_ERROR: too many SQL variables`.

For now I wrote simple chunk method, but I think it should be handled by strapi or bookshelf (they closed this issue).
